### ISSUE TITLE
Coding rule for model equations

### ISSUE THREAD

**Issue by jdebacker** [open-source-economics/OG-USA#110]
As we add more complexity to the model (in particular taxes on dynamically optimizing firms), I'm noticing that equations become quite messy.  This is leading to more algebra and coding errors on my part when I try to solve for an endogenous variable analytically.  

We should think about how to best balance algebraic simplification with computational efficiency?  I'm thinking of cases where you might be able to find a closed form solution, but it's very messy.  Would it be better to code up the necessary condition from the model and let a numerical root finder get the answer? The trade off seems to be computational time (the time to solve with a root finder vs the time to do simple calculations) versus accuracy (and flexibility) of the code.  

Basically, for our sake as we code this up initially (and for the ability to make the model more flexible in the future), it would be nice to to have something like in Dynare- where the user can just change/enter the necessary equations as specified by theory and off it goes.  This is in contrast to what I'm doing now where I do a lot of algebra to derive other equations from these necessary conditions, e.g. the factor demands of each firm.

**Comment by kerkphil** [comment]
This is the approach we take (have to take, I think) on the household side.  We have a series of intertemporal Euler equations and budget constraints for each household that must hold over their lifetime.  We have a constraint that the asseet holdings left at the end of some final period (age S+1 if there are no bequests) must be zero.  If we guess at the initial consumption value the Euler equations and budget constraints give us a whole time series of consumptions and asset holdings.  We use a root finder to get the value of initial consumption that sets final asset holdings to zero exactly.

This is much too complex a function to write out algebraically and solve.  I believe, in principle, we could do the same thing with profit-maximizing conditions for the firm.  it is probably more efficient to do so, as you rightly point out.

The devil is in the details, I suppose.  But it is a good idea to pursue.

**Comment by jdebacker** [comment]
Closing for lack of interest.
